### ISSUE TITLE
Improving the printout of the odML objects

### ISSUE THREAD

**Issue by lzehl** [G-Node/python-odml#67]
I think with the new updates of odML, an additional nice feature would be if the odML objects when printed provide more information of the underlying structure.

For example an odML.Document when printed gives currently the following information::
    >>> print(doc)
    >>> <Doc version author (number of directly attached sections)>
I think it would be useful if the odML.Document printout also indicates the depth (sections) of the tree structure and maybe the total number of properties? For example::
    >>> <Doc version author (# sections, depth #, grouping # properties)>

Same goes for the odML.Section printout. Here, besides the numbers of directly attached subsections also the direct number of properties, as well as the possible depth of the branch of the tree structure would be helpful.

One could even provide an extended print function which could summarize the content of the object in more detail (as a second option comparable to the XML structure).
For example::
    >>> print(doc, show_sec=True, show_prop=True)
    >>> <Doc version author>
               <Sec secname sectype>
                   <Prop propname dtype>         

Let me know what you think. Best Lyuba

**Comment by digaru19** [comment]
@lzehl  If this issue is not being worked on, I would like to work on this.
Can I take up this issue ?

**Comment by lzehl** [comment]
@digaru: sure :) feel free to assign yourself to this issue. I'm not working on it, and I'm glad if someone does. But what to print should maybe discussed with more people?

**Comment by lzehl** [comment]
Sorry I wrote you wrong... previous comment was ment for @digaru19

**Comment by mohit-0212** [comment]
@digaru19 I had already started working on the issue, and am almost through with it. I apologize for not updating this thread.

**Comment by digaru19** [comment]
@lzehl  Sure, we should discuss it with more people :v: .
 I feel that a minor issue can arise due the extended functionalities being asked for :- 
- The depths of all the children sections or sub-sections for a particular parent (document or a section, respectively),  may not be equal. So, we can't have a `depth` value attached to a `odml.Document` directly (unless we keep the `depth` as the maximum `depth` of its corresponding children) 

Your thoughts ?

**Comment by lzehl** [comment]
@digaru19: of course I meant the maximum depth of sections. If depth would be printed that is. Another possibility is just providing the number of directly attached and the total number of sections. Or all of it... I think we should just provide with the printout of an odml object a better overview of what it contains

**Comment by achilleas-k** [comment]
I'm curious whether there is the possibility of an issue with very large structures. I guess it depends on one's expectations, but it might be hard to work with a very large tree if printing triggers a full traversal. I'd be more comfortable with plain `print` just showing the number of directly attached children (for both the document and individual sections) and perhaps an extended print could be called explicitly which would force a full traversal.

I'd be interested to hear whether my concern is valid though. Is it possible for a full traversal to take long for very large documents?

**Comment by digaru19** [comment]
@achilleas-k  Yeah, a full traversal for very large documents may take a significant amount of time. And considering that the user is printing the full traversal on a regular console, it might exceed the scrollable view of the console, and the user won't be able to view the complete output. 
  I will go by your views, that ultimately it will depend on one's expectations and choices. Let the current implementation for `print` be as such, and have an `extended print` kind of function, which can force a full traversal, but it won't print the whole tree structure as such, rather it will print a minified summary for each of the components of the tree.  

 This way, both of our concerns can be satisfied to a great extent, i.e. 
- We get a better overview of the whole tree structure.
- Though, it will force a full traversal, but it can be controlled by having custom parameters in the `extended print` function. 

@achilleas-k  @lzehl  Thoughts ??

**Comment by achilleas-k** [comment]
Well I didn't mean to imply that a `print` would display the full tree, just that you would need to traverse it to count the number of children and the max depth.

Anyway, I agree with the separation and your description. I'd wait for @lzehl to sign off on the idea though.

**Comment by lzehl** [comment]
@digaru19, @achilleas-k and @jgrewe:

*time of a full traversal*: 
In my experience with the iter functions a full traversal of sections of our tree size which is rather big and complex did not take that much time. It could though when including properties and especially values. My conclusion: it could become inconvenient at one point.

Would it be possible to *store* the total number of child objects automatically and gradually? Especially for a document?

**Comment by lzehl** [comment]
@digaru19, @achilleas-k, @jgrewe:

My problem with the current print is, that it is misleading. If a document print states "4 sections" most people (incl me in the beginning) will assume that the document consists of 4 sections in total...

If we want to keep it simple I would suggest therefore to rather kick this out and provide other helper functions that help the user to get familiar with the struture of an odml.

**Comment by jgrewe** [comment]
not sure, whether storing is feasible. This would require to update every time a section is appended or, even worse, a tree is added to a just-created document for saving. Then, a full traversal would be required for a feature that might not be wanted at that particular moment. I would probably prefer to wait a short while when I request the information

**Comment by lzehl** [comment]
@digaru19, @achilleas-k, @jgrewe:

*printing a complete tree*: 
@jgrewe is correct that there is a function that prints the json and xml like representation of an odml tree, they are not that nice though. In general I must say that printing a complete tree is seldom usefull, printing a summarized tree struture though can be quite helpful when building or getting familiar with complex trees. In such a case the actual properties but especially the values are rather uninteresting and could be summarized or left out.

**Comment by lzehl** [comment]
@jgrewe: I agree this would require updates with adding or removing objects, but a full traversal would not be necessary if I'm not wrong even if complete tree branches are attached. It would  then require a summation of the stored values of the parent and the attached child object. For example, a document consists in total of 10 sections. A new section with a branch of 5 subsections are now attached to this document. The resulting document would then contain 10+1+5=16 sections... but I'm not sure how complicated the implementation of such a functionality would be?

**Comment by digaru19** [comment]
@lzehl I definitely agree with what you have mentioned.
Also, a small addition to what @jgrewe had mentioned earlier, regarding the full traversal on appending a tree to a document. Just like a document will have `total sections`, `depth`, etc., we can have can have similar things for a `section`. So, when we are building a tree, all the sections in it will hold the required information, and will be updated on each addition or deletion operation, and thus, when adding it to a newly created doc, it won't lead to a full traversal, but will be just a single operation. 

Having such information attached individually to each section can also allow us to add other helpful features in the future too. 

Though, if we are following this approach, the update operations will take longer (Order of tree height), and will be the costliest on addition or deletion of a leaf node in a very huge tree structure.

**Comment by achilleas-k** [comment]
I was just about to mention the same. Any change at a deep node would require updating the counts all the way up the tree.

**Comment by jgrewe** [comment]
I just wondered how ````du```` is working and actually you pass an optional depth argument. By default a detailed view for each node is given. It seems to traverse the tree. 
Inodes are ignorant of their size, btw. Maybe for a good reason?

**Comment by achilleas-k** [comment]
`du` visits all files regardless of depth argument. Specifying it simply displays the size of everything under any directory beyond that depth.

**Comment by msuzen** [comment]
:+1: Nice discussion. 

Direction goes into more like `odML.search` or `odml.summary` functionality,  far beyond pretty printing.  Maybe some existing APIs can be leveraged like [xml.etree](https://docs.python.org/2/library/xml.etree.elementtree.html#module-xml.etree.ElementTree) or [xml.dom](https://docs.python.org/2/library/xml.dom.html) in building this kind of functionality, if `odml` object can be converted to DOM, before writing it into file.

**Comment by mohit-0212** [comment]
Interesting Discussion. I just made a recursive function that calculated the total no of sections/subsections and total no of properties and returned it, everytime user asked for it, but this discussion delves into a much more interesting functionality where @lzehl talks about printing a summarized tree structure for a document. Surely this would give a better idea of how the doc is organized. But there might be some time constraints considering a very large document, more so in case of an updation (which can happen very frequently, so resulting in that amt of time on every updation), also knowing that the length of a document is not limited
Agree with @msuzen where he suggests the use of already existing APIs for building such a tree.

**Comment by lzehl** [comment]
Good morning everyone! This issue turns out to be more interesting and complicated than I  originally expected :)

I think I need a summary:

1. **updating stats of structure for objects**: idea is out, because of the potential traversal updates

2. **# sections in current print**: I would like to kick this info out, because it's misleading. I'm not sure, but maybe renaming this into # subsection could make it a bit clearer especially if for section objects also the # properties is given. What do you think?

3. **total depth, total # sections etc**: should be implemented as separate functions?

4. **printing tree structure summaries**: is nice to have but should go into separate functions and should maybe profite from available APIs. 

5. Btw this could be a related topic for point 4: it would be nice to provide a visualization of the trees (or tree parts or summaries) as well. This is a bigger issue though. Optimal XML visualization is not that easy. I started to try out some available packages for this (poster on metadata orga with odml for INCF conf in Leiden, NL), but did not derive a good solution especially for bigger trees (which could be due to programming skills in my case). Any ideas?

**Comment by achilleas-k** [comment]
Morning!

Seems like we all agree that anything that requires **full traversal** of the tree should be made when the information is requested explicitly (total depth, total section count).

So now all we need is to figure out what useful information we want to print in the simpler case. This might be unusual, but how do you feel about something like the following:
```python
>>> print(doc)
<Doc VER by Author (2 children, unknown descendants)>
>>> doc.update_tree_info()  # traverses tree and calculates all information
>>> print(doc)
<Doc VER by Author (2 children, 15 descendants, depth 4)>
>>> sec = odml.Section(...)  # new section
>>> doc.append(sec)
>>> print(doc)
<Doc VER by Author (2 children, unknown descendants)>
```

This is remix of the ideas discussed so far.
- Instead of having an extended `print` function, the normal printing displays as much information as is available.
- At any given time, the number of children (directly attached Sections) is available. By default, the total number of descendants is listed as `unknown`. This maybe gets around the issue of the number of sections printed being misleading information, since we can use two different terms to distinguish between directly attached subsections and total subsections.
- If the `update_tree_info()` method is called, a full traversal is triggered. When the `print` is called again, the calculated info is shown.
- The calculated info is invalidated as soon as a change is made.

Obviously the terms `children` and `descendants` can be changed to odML specific terms (I am using tree data structure terms here).

It maybe tempting to think we could invalidate only the sections above where the change is made, because in the example the number of descendants of a section at depth 2 doesn't change, but then we're back to the same idea of having to propagate changes up the tree. By having a global `extended_info_valid` flag on the root of the document, we only perform one additional operation when the tree structure changes.

**Comment by lzehl** [comment]
@achilleas-k: I like this approach with odml specifi

**Comment by lzehl** [comment]
Oh nooooo, how do I reopen the issue? I'm writing feom my cell.... and closed this by accident...

**Comment by achilleas-k** [comment]
Reopened :)

**Comment by lzehl** [comment]
I'm so sorry... ;) 

What I meant to comment was: I like this approach with odml specific terms. # sections for doc and sec objects and for latter additionally # properties. I like the term descendants and would stick there just to the descending sections.

**Comment by jgrewe** [comment]
I like the approach! The only thing I see is how to make sure, that the information related to the doc is accurate and up-to-date.

**Comment by achilleas-k** [comment]
@lzehl So just to clarify, based on your comments, the example becomes:
```python
>>> print(doc)
<Doc VER by Author (2 sections, unknown descendants)>
>>> print(doc.sections["SectionOne"]
<Section SectionOne[sectype] 1 section, 3 properties, unknown descendants>
>>> doc.update_tree_info()  # traverses tree and calculates all information
>>> print(doc)
<Doc VER by Author (2 sections, 15 descendants, depth 4)>
>>> sec = odml.Section(...)  # new section
>>> doc.append(sec)
>>> print(doc)
<Doc VER by Author (3 sections, unknown descendants)>
```

@jgrewe I think the most straightforward approach is to set `doc.extended_info_valid = True` at the end of the `doc.update_tree_info()` method and then have a `doc.invalidate_extended_info()` (or just set to `False`) in **all** append methods.
There's an issue with this, though: Sections have references to their parent but not the root of the tree, so calling `append` on a deep section again requires traversing up the tree to get to the root doc. Could we have each Section keep a reference to the root of the document the same way it keeps a reference to the parent? That way we would just have to do something like `self.root_doc.invalidate_extended_info()` at the end of each append or remove.

**Comment by achilleas-k** [comment]
On second thought, a reference to the root doc becomes a problem (again) when we append subtrees. It would require updating all Sections in the subtree to point to their new root.

**Comment by lzehl** [comment]
Yes that's exactly what I mean.

The `validate_extendet_info` and `invalidate_extendet_info` is needed for append and remove functions. For adding subtrees I think the again traversal updates are fungible. Since this only happens when the user explicitly calls the `update_tree_info` again. Correct?

**Comment by achilleas-k** [comment]
I think there's no getting around at least one operation requiring traversal and propagation of information. Here's another summary of everything that I'm thinking.

The `update_tree_info()` method will, of course, require parsing the entire tree. This is clear.
Beyond that, we either need to:
1. Keep a global flag that tells us whether the info is still valid.
2. Keep a per-section flag which indicates whether the information *below* it is valid.
3. Keep up to date information on each Section.

More details:
1. This requires the least processing (I think). It requires each Section to have a reference to the root object. Whenever an append or delete is performed, the Section invalidates the global information.
**Downside**: When appending N sections anywhere in the tree, we need to visit all N (newly appended) sections and update their reference to `root`. None of the existing sections in the tree are affected by this.
2. This allows us to partially invalidate information. When a Section is appended or deleted, the information on all sections above it needs to be invalidated. **Downside**: Appending or deleting a Section deep in the tree (at depth N), requires N updates. However, if the Section parent's flag is already invalid, we can stop the propagation, so it will only be an issue for the first append after an info update.
3. Keeping up to date information on each Section is similar to 2, but the downside of needing to propagate up the tree has no condition were we can stop. *Every* append or delete will require updating all Sections above. The upside is that `update_tree_info()` is no longer required, since we're constantly keeping information up to date.

3 is the worst by far. I thought 2 was pretty bad as well, but with the stop condition it becomes less of an issue. 1 has the advantage that the only operation which is affected is a multi-append operation, which on the other hand may be more common than a deep append or delete. Of course, the major disadvantage 1 has against the other options it that updating the information *always* goes through the entire tree, whereas with 2, it may often only require partial updates.

**Comment by achilleas-k** [comment]
I think I've convinced myself that @jgrewe is right to prefer 2. Trees don't become that deep anyway.

**Comment by mpsonntag** [comment]
Has been solved via the `pprint()` method in PRs #313 and #321.
